src/main/java/com/exasol/artifactreferencechecker/ArtifactReferenceCheckerMojo.java

```diff
@@ -0,0 +1,156 @@
+            DocumentBuilder builder = builderFactory.newDocumentBuilder();
+            Document xmlDocument = builder.parse(fileIS);
+            XPath xPath = XPathFactory.newInstance().newXPath();
+            String expression = "string(/project/build/plugins/plugin[artifactId=\"maven-assembly-plugin\"]/configuration/finalName)";
+            return (String) xPath.compile(expression).evaluate(xmlDocument, XPathConstants.STRING);
+        } catch (XPathExpressionException | IOException | SAXException | ParserConfigurationException exception) {
+            throw new IllegalStateException("Could not find finalName in maven-assembly-plugin");
+        }
+    }
+
+    private boolean matchPatternInProjectFiles(String regex, String expected) throws MojoExecutionException {
+        final Pattern pattern = Pattern.compile(regex);
+        final File projectDirectory = project.getModel().getProjectDirectory();
+        final FileValidationVisitor fileValidationVisitor = new FileValidationVisitor(getLog(), pattern, expected);
+        try {
+            Files.walkFileTree(projectDirectory.toPath(), fileValidationVisitor);
+        } catch (IOException exception) {
+            throw new MojoExecutionException("Could not check files. Cause: " + exception.getMessage(), exception);
+        }
+        return fileValidationVisitor.isSuccess();
+    }
+
+    /**
+     * File Visitor that checks of the files contain outdated artifact references
+     */
+    private static class FileValidationVisitor extends SimpleFileVisitor<Path> {
+        private final Log log;
+        private final Pattern pattern;
+        private final String expected;
+        List<String> extensions = List.of("java", "md");
+        private boolean success = true;
+
+        private FileValidationVisitor(Log log, Pattern pattern, String expected) {
+            this.log = log;
+            this.pattern = pattern;
+            this.expected = expected;
+        }
+
+        private boolean hasCorrectEnding(Path path) {
+            return extensions.stream().anyMatch(extension -> path.toString().endsWith("." + extension));
+        }
+
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+            if (hasCorrectEnding(file)) {
+                validateFile(file);
+            }
+            return FileVisitResult.CONTINUE;
+        }
+
+        private void validateFile(Path file) throws IOException {
+            try (Scanner fileScanner = new Scanner(file)) {
+                while (fileScanner.hasNextLine()) {
+                    String line = fileScanner.nextLine();
+                    validateLine(file, line);
+                }
+            } catch (FileNotFoundException exception) {
+                throw new IllegalStateException(
+                        "Could not open project file " + file.toString() + ". Cause: " + exception.getMessage(),
```

Comment:
I think, you can remove the cause. It should propagate with exception.

Usually, cause is required in our udfs.